notifications/notifications.go

```diff
@@ -20,42 +20,35 @@ type PubSub interface {
 func New() PubSub {
 	return &impl{
 		wrapped: *pubsub.New(bufferSize),
-		cancel:  make(chan struct{}),
 	}
 }
 
 type impl struct {
+	lk      sync.RWMutex
 	wrapped pubsub.PubSub
 
-	// These two fields make up a shutdown "lock".
-	// We need them as calling, e.g., `Unsubscribe` after calling `Shutdown`
-	// blocks forever and fixing this in pubsub would be rather invasive.
-	cancel chan struct{}
-	wg     sync.WaitGroup
+	closed bool
 }
 
 func (ps *impl) Publish(block blocks.Block) {
-	ps.wg.Add(1)
-	defer ps.wg.Done()
-
-	select {
-	case <-ps.cancel:
-		// Already shutdown, bail.
+	ps.lk.RLock()
+	defer ps.lk.RUnlock()
+	if ps.closed {
 		return
-	default:
 	}
 
 	ps.wrapped.Pub(block, block.Cid().KeyString())
 }
 
 // Not safe to call more than once.
```

Comment:
Pretty sure this comment no longer relevant. I mean not that you should, but looks like it'll just return early if you do